lib/tests/robustness.js

```diff
@@ -59,6 +59,29 @@ function assertDeterministicTypes(node, path = '') {
         assertDeterministicTypes(node.items, `${path}/items`);
     }
 
+    // if node type is object, assert it has a schema or is a map
+    if (node.type == 'object') {
+        const hasSchema = node.properties || node.oneOf || node.allOf;
+        if (!hasSchema && !node.additionalProperties) {
+            throw new assert.AssertionError({
+                message: `object type must specify schema or additionalProperties: #/${path}`
+            });
+        }
+
+        // Object types can't have additionalProperties
+        if (hasSchema && node.additionalProperties) {
```

Comment:
It might be possible to use `properties` + `additionalProperties`, if `additionalProperties` is a valid map type.  You could use `properties` to manually define a few of the 'required' keys, but then allow for additional unspecified keys.  

But, this maybe could also be done using [propertyNames](https://json-schema.org/understanding-json-schema/reference/object.html#property-names) ? 

Anyway, I'm fine with this restriction, we should probably run this test on WMF schema repos to make sure it doesn't break any existing uses.

Reply:
Ran it on primary, and it does break a few schemas so I guess we can drop the test:
`mediawiki/recentchange`
`mediawiki/revision/create`
`w3c/reportingapi/network_error`

More importantly, the other test on object types specifying anything at all fails on a few schemas:
`change-prop/continue`
`change-prop/retry`
`mediawiki/job`
`mediawiki/page/properties-change`

Should we add these to the ignore list for deterministic types?

Reply:
> we can drop the test
Okay, sounds good.

> Should we add these to the ignore list for deterministic types
Yes, we should.

Reply:
PR for primary schemas:
https://gerrit.wikimedia.org/r/c/schemas/event/primary/+/929782

All the secondary schemas seem to pass the tests